File: service/app/src/main/java/io/miragon/blueprint/application/service/GetPendingClarificationsService.java
package io.miragon.blueprint.application.service;

import io.miragon.blueprint.application.port.inbound.GetPendingClarificationsQuery;
import io.miragon.blueprint.application.port.outbound.BikePortfolioRepository;
import io.miragon.blueprint.application.port.outbound.LeasingApplicationRepository;
import io.miragon.blueprint.application.port.outbound.TaskInboxPort;
import io.miragon.blueprint.domain.bike.Bike;
import io.miragon.blueprint.domain.bike.BikeId;
import io.miragon.blueprint.domain.leasing.LeasingApplication;
import io.miragon.blueprint.domain.leasing.PendingClarification;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional(readOnly = true)
public class GetPendingClarificationsService implements GetPendingClarificationsQuery {

    private final TaskInboxPort taskInbox;
    private final LeasingApplicationRepository repository;
    private final BikePortfolioRepository bikePortfolio;

    public GetPendingClarificationsService(
            TaskInboxPort taskInbox,
            LeasingApplicationRepository repository,
            BikePortfolioRepository bikePortfolio) {
        this.taskInbox = taskInbox;
        this.repository = repository;
        this.bikePortfolio = bikePortfolio;
    }

    @Override
    public List<PendingClarification> pending() {
        List<TaskInboxPort.OpenClarification> openTasks = taskInbox.findOpenClarifications();
        // Pair each open task with its application; skip tasks whose application vanished (defensive).
        List<Map.Entry<TaskInboxPort.OpenClarification, LeasingApplication>> cases = openTasks.stream()
                .flatMap(task -> repository.findById(task.applicationId())
                        .map(application -> Map.entry(task, application))
                        .stream())
                .toList();
        Map<BikeId, String> models = new HashMap<>();
        for (Bike bike : bikePortfolio.findAllByIds(
                cases.stream().map(c -> c.getValue().bikeId()).toList())) {
            models.put(bike.bikeId(), bike.model());
        }
        return cases.stream()
                .map(c -> new PendingClarification(
                        c.getValue().id(),
                        c.getValue().customerName(),
                        c.getValue().bikeId(),
                        models.get(c.getValue().bikeId()),
                        c.getKey().waitingSince()))
                .toList();
    }
}

File: service/app/src/main/kotlin/io/miragon/blueprint/application/service/GetPendingClarificationsService.kt
package io.miragon.blueprint.application.service

import io.miragon.blueprint.application.port.inbound.GetPendingClarificationsQuery
import io.miragon.blueprint.application.port.outbound.BikePortfolioRepository
import io.miragon.blueprint.application.port.outbound.LeasingApplicationRepository
import io.miragon.blueprint.application.port.outbound.TaskInboxPort
import io.miragon.blueprint.domain.leasing.PendingClarification
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class GetPendingClarificationsService(
    private val taskInbox: TaskInboxPort,
    private val repository: LeasingApplicationRepository,
    private val bikePortfolio: BikePortfolioRepository,
) : GetPendingClarificationsQuery {

    override fun pending(): List<PendingClarification> {
        val openTasks = taskInbox.findOpenClarifications()
        // Pair each open task with its application; skip tasks whose application vanished (defensive).
        val cases = openTasks.mapNotNull { task ->
            repository.findById(task.applicationId)?.let { application -> task to application }
        }
        val models = bikePortfolio
            .findAllByIds(cases.map { (_, application) -> application.bikeId })
            .associate { it.bikeId to it.model }
        return cases.map { (task, application) ->
            PendingClarification(
                applicationId = application.id,
                customerName = application.customerName,
                requestedBikeId = application.bikeId,
                requestedBikeModel = models[application.bikeId],
                waitingSince = task.waitingSince,
            )
        }
    }
}
